### ISSUE TITLE
[Feature request] Versioning or Save-As

### ISSUE THREAD

**Issue by zigmhount** [Houston4444/RaySession#45]
Hello,
First thank you so much for this nice piece of software :)

In many projects it is critical to be able to compare versions, e.g. after changing synth parameters or effects or EQ... Ray has the option to duplicate an existing project to give it another name or version number, but it has to be done *before* making the changes. In many cases however I (but maybe that's only me!) don't know yet what I will be changing until I do, and saving my current changes with a new session name would be much easier.

This could be implemented either with a Save As command, and/or (better IMO) save as a new version (LMMS and Ardour have that) simply adding an incremented integer to the session's name.
I guess that it would require to make first a temporary copy of the whole folder before starting to work on it, which might make the intermediate saving of applications settings/files within the project less flexible...

Alternatively it would already make things easier if there was a button to create a new version of an existing project when selecting a project to open.

What do you think?

**Comment by Houston4444** [comment]
Hi !
Are you sure that the feature you want isn't already done by snapshots ?
between the "Executable" button and the server status line you've got a button with a "fast forward" icon.
If you don't have this icon, it means you have not 'git' installed.
To install git on debian like systems, just run 'sudo apt install git'.
Then, it control menu, enable 'auto snapshots'.

With this, each time the session is saved, a snapshot is taken. And you can easily go back to a previous snapshot clicking on the "fast-forward" button.

**Comment by zigmhount** [comment]
Aaaah right, it sounds like it would cover exactly what I ask! I should indeed have tested the latest release before raising a request, I'm on version 0.7.2 on LibraZik2, sorry for the noise. I'll upgrade and confirm here once I've checked that feature!

**Comment by zigmhount** [comment]
Hmm, after reading snapshots_explain I don't think it's exactly what I had in mind.
See my user case: I open a session, make some changes and save them (let's call that version 1.0), make some other changes but I don't save yet, decide to replace a synth with another and some EQ/compressor tweaks, and save (version 1.1). The next day I'd like to listen both versions again to make sure it was a good choice (or not).

If I understand right, if I open the last save (v1.1) and rewind to snapshot 1.0, v1.1 would be discarded, correct?
So I would have to open the last save (v1.1), duplicate it with another name (e.g. v2.0), open again the last save (v1.1), and rewind to the previous snapshot (v1.0), which I could then listen to again, but it would discard v1.1. Am I right?
I would like instead to be able to chose the snapshot v1.0, listen to it, (maybe duplicate it into a new version) and go back to snapshot v1.1. Not sure whether what I'm writing makes sense, can you follow?

Side question: if I duplicate a session after having made changes in it, will it first save all changes in the "old" session before duplicating? (if it doesn't, it's pretty much what I'm looking for).

I suppose that what I want may be doable with the same system as currently used, but I don't know Git (apart of git clone, pull or checkout) so I'd have to look into some documentation to properly understand.

**Comment by Houston4444** [comment]
> If I understand right, if I open the last save (v1.1) and rewind to snapshot 1.0, v1.1 would be discarded, correct?
No, you will be able to go back to v1.1 without any problem.

> Side question: if I duplicate a session after having made changes in it, will it first save all changes in the "old" session before duplicating?
No, session is saved before to be duplicated. It would be not sane to not save before for many reasons (we can't know if clients were saved from themselves, temp files created by clients not removed...).

**Comment by zigmhount** [comment]
Makes sense, thanks for your answer! I'll close this ticket and start using snapshots then :)

**Comment by Houston4444** [comment]
Note that, for the use case you talk about, choosing between 2 patches for an instrument, I would prefer different workflows:

- put an A/B switch plugin somewhere to can switch sound very quickly (it uses more CPU, but it allows a direct comparaison).
- save the needed client as template, stop it, start another client with its template, modify it, and compare the 2 clients switching which client you want to be started.

Snapshots goal is IMO most for theses cases

-  You feel to just have done bullshit modifications the last moments because you were really not inspired finally
- You added a plugin somewhere making the application crash, and you want to go back the moment it was working well

**Comment by zigmhount** [comment]
Very interesting, thanks so much for the suggestions!
